### DISCUSSION TITLE
Leaked types and shared_ptr

### DISCUSSION THREAD

**Discussion by chrisrichardson** [wjakob/nanobind#353] [Q&A] [answered]
I'm trying to understand why I get a lot of `leaked type` messages on exit from nanobind. I have some code that looks a bit like this:

```
#include <nanobind/nanobind.h>
#include <nanobind/stl/shared_ptr.h>

namespace nb = nanobind;

class Owl
{
public:
  Owl(){};
};

class Barn
{
public:
  Barn(std::shared_ptr<Owl> owl) : _owl(owl){};

  std::shared_ptr<Owl> _owl;
};

NB_MODULE(example, m)
{

  nb::class_<Owl>(m, "Owl").def(nb::init<>());

  nb::class_<Barn>(m, "Barn")
      .def(nb::init<std::shared_ptr<Owl>>())
      .def_ro("owl", &Barn::_owl);
}        
```

Now doing in Python:
```
owl = Owl()

w = Barn(owl)
```
will trigger a leaked type warning. 
```
 - leaked type "example.Owl"
```

Looking at `nanobind/stl/shared_ptr.h` it seems like `py_deleter` is being called on `Owl` **after** the interpreter has exited, so the refcount is not being decremented. Is this expected behaviour? Is there anything I can do better to prevent it?

**Comment by wjakob** [comment, marked as answer]
Older versions of Python are sometimes a bit leaky especially on the global scope. Does the issue happen if you wrap
```
owl = Owl()

w = Barn(owl)
```
into a function and then call that function?

**Reply by wjakob** [reply]
You can disable the leak warning by the way, refer to the documentation for details. In pybind11, you simply did not know about leaks, so it's a benefit for developers.

**Reply by chrisrichardson** [reply]
Yes, wrapping in a function gets rid of the problem. Maybe it will be better with Python 3.12 (I have 3.10 on this machine).
I know we can disable warnings (and have done it for Release builds)... 
Thanks for the quick reply.
